libs/wire-api-federation/test/Test/Wire/API/Federation/ClientSpec.hs

```diff
@@ -0,0 +1,148 @@
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# OPTIONS_GHC -Wno-deferred-type-errors #-}
+{-# OPTIONS_GHC -Wno-partial-type-signatures #-}
+
+module Test.Wire.API.Federation.ClientSpec where
+
+import qualified Control.Concurrent.Async as Async
+import Control.Exception (bracket, finally, throwIO, try)
+import Control.Monad.Except (ExceptT, MonadError (..), runExceptT)
+import Control.Monad.State (MonadState (..), modify)
+import Control.Retry (constantDelay, limitRetries, retrying)
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString.Lazy as LBS
+import Data.Domain (Domain (Domain))
+import Foreign.C (Errno (Errno), eCONNREFUSED)
+import GHC.IO.Exception (ioe_errno)
+import Imports
+import Mu.GRpc.Client.Record (grpcClientConfigSimple)
+import Mu.GRpc.Server (msgProtoBuf, runGRpcAppTrans)
+import Mu.Server (MonadServer, ServerError, ServerErrorIO, SingleServerT)
+import qualified Mu.Server as Mu
+import Network.Socket (Family (AF_INET), SockAddr (SockAddrInet), SocketType (Stream), close', connect, socket, tupleToHostAddress)
+import Test.Hspec
+import Test.QuickCheck (arbitrary, generate)
+import qualified Wire.API.Federation.API.Brig as Brig
+import Wire.API.Federation.Client (FederationClientError (FederationClientOutwardError, FederationClientRPCError), FederatorClient, runFederatorClientWith)
+import Wire.API.Federation.GRPC.Client (createGrpcClient)
+import Wire.API.Federation.GRPC.Types (Component (Brig), FederatedRequest (FederatedRequest), Outward, OutwardError, OutwardResponse (OutwardResponseBody, OutwardResponseError), Request (..))
+import Wire.API.User (UserProfile)
+
+spec :: Spec
+spec = fdescribe "Federator.Client" $ do
+  it "should make correct calls to the federator and parse success response correctly" $ do
+    handle <- generate arbitrary
+    expectedResponse :: Maybe UserProfile <- generate arbitrary
+
+    (actualResponse, sentRequests) <-
+      withMockFederator (mkSuccessResponse expectedResponse) $
+        Brig.getUserByHandle Brig.clientRoutes handle
+
+    sentRequests `shouldBe` [FederatedRequest "target.example.com" (Just $ Request Brig "/federation/users/by-handle" (LBS.toStrict (Aeson.encode handle)) "origin.example.com")]
+    actualResponse `shouldBe` Right expectedResponse
+
+  it "should parse failure response correctly" $ do
+    handle <- generate arbitrary
+    someErr <- generate arbitrary
+
+    (actualResponse, _) <-
+      withMockFederator (mkErrorResponse someErr) $
+        Brig.getUserByHandle Brig.clientRoutes handle
+
+    actualResponse `shouldBe` Left (FederationClientOutwardError someErr)
+
+  it "should report federator failures correctly" $ do
+    handle <- generate arbitrary
+
+    (actualResponse, _) <-
+      withMockFederator (error "some IO error!") $
+        Brig.getUserByHandle Brig.clientRoutes handle
+
+    -- This behaviour is actually incorrect, the fix in upstream is not merged
+    -- yet: https://github.com/haskell-grpc-native/http2-grpc-haskell/pull/48
```

Comment:
Is it possible to depend on the fork with the fix rather than broken upstream?

Reply:
Done!